src/lang-promql/complete.ts

```diff
@@ -0,0 +1,117 @@
+// Every different completion mode must implement this interface
+interface Complete {
+  promQL(context: AutocompleteContext): Promise<CompletionResult> | CompletionResult | null;
+}
+
+// LSPComplete will proviode an autocompletion based on what the langserver-promql is providing when requested
+class LSPComplete implements Complete {
+  private readonly lspClient: LSPClient;
+
+  constructor(lspClient: LSPClient) {
+    this.lspClient = lspClient;
+  }
+
+  promQL(context: AutocompleteContext): Promise<CompletionResult> {
+    const {state, pos} = context
+    const line = state.doc.lineAt(pos)
+    const body: LSPBody = {
+      expr: state.doc.sliceString(0),
+      limit: 100,
+      // positionChar start at 0 at the beginning of the line for LSP
+      positionChar: pos - line.from - 1 > 0 ? pos - line.from - 1 : pos - line.from,
+      // positionLine start at 0 from LSP and at 1 from CMN
+      positionLine: line.number - 1,
+    }
+    return this.lspClient.complete(body)
+      .then((items: CompletionItem[]) => {
+        const options: Completion[] = []
+        // for every `textEdit` present, they all have exactly the same `range.start` value.
+        // so the goal is to save the last one present in order to use it later to calculate the position `from`
+        // Note: textEdit can be undefined but that's just because otherwise it doesn't compile
+        // see more here: https://github.com/microsoft/TypeScript/issues/12916
+        // and here https://github.com/microsoft/TypeScript/issues/9568
+        let textEdit: TextEdit | undefined
+
+        if (isIterable(items)) {
+          for (const res of items) {
+            // TODO map kind with icon Completion.type when it is released on CMN side
+            // https://github.com/codemirror/codemirror.next/commit/459bba29c1fd1d80fc4f36dac27e5825b2362273
+
+            // `apply` is the string that will be used when the completion is performed.
+            // By default it's the same value as the label (i.e the same string as the one shown in the completionList)
+            let apply = res.label
+            if (res.textEdit) {
+              apply = res.textEdit.newText
+              textEdit = res.textEdit
+            }
+            options.push({label: res.label, apply: apply})
+          }
+        }
+
+        // `from` and `to` are the absolute value in term of character and doesn't consider the line number
+        // so we have to calculate the position of "from"
+        // `to` is the direct value of pos
+        return {
+          from: textEdit ? line.from + textEdit.range.start.character : line.from,
+          to: pos,
+          options: options
+        } as CompletionResult
+      })
+  }
+}
+
+// OfflineComplete is going to provide a full completion result without any distant server
+// So it will basically only provide the different keyword of the PromQL syntax
```

Comment:
There might also be a use case for an autocompletion of metric names that doesn't go back and forth to the server for every keystroke? Since the metric names can be loaded once upon page load, and then you can do nice and quick fuzzy matching against them, and the grammar tells us when we are completing a metric name.

Reply:
yes I'm thinking also of an hybrid mode let's say. Since LSP will be available in a future version of Prometheus, then we should have a mode when LSP is not available, to be a bit retro compatible with the different version of prometheus

Reply:
Having a hybrid mode sounds hard to get right, since it's unclear, when the server should be called. The completions that are most expensive to compute are metric names and labels and that's something a offline mode can't do anyway.

What seems more useful to me is caching API responses, as long as the user is typing the same word.

Reply:
caching is just an optimization. It will be handled by a `prometheusClient` or by the hybidMode itself. 

For example like that: 

```typescript
class HybridComplete implements Complete {
  private metrics: string[];
  promQL(context: AutocompleteContext): Promise<CompletionResult> {
      const {state, pos} = context
     // Analyze the tree.
     // When required call the prometheusClient
     if (metrics.lenght) <= 0 {
         this.promClient.getMetricName()
           .then((result) => {
             this.metrics = result
             // tranform the result to the correct format
           })
     }
  }
}

Reply:
I think this mode should able to replace properly what you are doing with Monaco @juliusv

Reply:
@Nexucis To fully replace what I get "out of the box" with Monaco, I would need:

* Fast offline fuzzy matching against once-loaded metric names (good, with proper scoring, etc.) (https://github.com/codemirror/codemirror.next/issues/236)
* Highlighting of partial fuzzy match results (https://github.com/codemirror/codemirror.next/issues/234)
* Icons that indicate the completion item type (https://github.com/codemirror/codemirror.next/issues/226)

Reply:
> caching is just an optimization. It will be handled by a prometheusClient or by the hybidMode itself.

I was talking about caching the langserver API responses and then applying offline fuzzy matching. That is nothing that is specific to metric name completions.

In fact, when using the language server API, CM doesn't need to be concerned with what kind of completion to look for.

Reply:
I'm still not sure what problems a hybrid mode, that does more than caching and fuzzy matching, would solve.

Reply:
Yeah I know don't worry. I was more talking about what you have with the auto-completion in term of data. Rendering and visualization is another part.
* For the icons, it's just matter of time since it's already implemented, we just have to wait the next release of codemirror.
* For the fuzzy matching, well as you can see I'm waiting the answer of marijnh. So I didn't have better idea, and it should be enough to cover the needs IMHO.
* For the hightlighting, I'm still searching how to do it. I think that's the part that will take me the most time. Others should be quite easier, just matter of algorithm and looping on struct ^^

Reply:
> I'm still not sure what problems a hybrid mode, that does more than caching and fuzzy matching, would solve.

It will be a short replacement of LSP when it's not available. I'm saying a "short" replacement, because for sure it won't be possible/easier to implement all functionnality of the completion provided by LSP.